ostd/src/sync/rcu/mod.rs

```diff
@@ -394,32 +398,51 @@ impl<P: NonNullPtr + Send> RcuOptionReadGuard<'_, P> {
     }
 }
 
-/// # Safety
+/// A wrapper to delay calling destructor of `T` after the RCU grace period.
+///
+/// Upon dropping this structure, a callback will be registered to the global
+/// RCU monitor. And the destructor of `T` will not be immediately called.
 ///
-/// The pointer must be previously returned by `into_raw` and the pointer
-/// must be only be dropped once.
-unsafe fn delay_drop<P: NonNullPtr + Send>(pointer: NonNull<<P as NonNullPtr>::Target>) {
-    struct ForceSend<P: NonNullPtr + Send>(NonNull<<P as NonNullPtr>::Target>);
-    // SAFETY: Sending a raw pointer to another task is safe as long as
-    // the pointer access in another task is safe (guaranteed by the trait
-    // bound `P: Send`).
-    unsafe impl<P: NonNullPtr + Send> Send for ForceSend<P> {}
+/// `RcuDrop<T>` is guaranteed to have the same layout as `T`. You can also
+/// access the inner value safely via `RcuDrop<T>`.
+#[derive(Clone, Debug, Default, PartialEq, Eq, PartialOrd, Ord, Hash)]
+#[repr(transparent)]
+pub struct RcuDrop<T: Send + 'static> {
+    value: ManuallyDrop<T>,
+}
 
-    let pointer: ForceSend<P> = ForceSend(pointer);
+impl<T: Send + 'static> RcuDrop<T> {
+    /// Creates a new [`RcuDrop`] instance that delays the dropping of `value`.
+    pub fn new(value: T) -> Self {
+        Self {
+            value: ManuallyDrop::new(value),
+        }
+    }
+}
 
-    let rcu_monitor = RCU_MONITOR.get().unwrap();
-    rcu_monitor.after_grace_period(move || {
-        // This is necessary to make the Rust compiler to move the entire
-        // `ForceSend` structure into the closure.
-        let pointer = pointer;
+impl<T: Send + 'static> Deref for RcuDrop<T> {
+    type Target = T;
 
-        // SAFETY:
-        // 1. The pointer was previously returned by `into_raw`.
-        // 2. The pointer won't be used anymore since the grace period has
-        //    finished and this is the only time the pointer gets dropped.
```

Comment:
It is clever to implement `delay_drop` with `RcuDrop`. But the current implementation is not sound. This is a `frow_raw` security requirement that exists for a reason, because Rust enforces `noalias` on the magic `Box` type. Now that requirement has been deleted, and in fact it is not met. So the code won't be sound as well.

So even if you want to use `RcuDrop` in `delay_drop`, you have to keep the wrapper type and do `from_raw` in the `drop` method, which is fine because by the time the `drop` method is called, the RCU grace periods have passed, so there is no other use for the pointer.